sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java

```diff
@@ -622,20 +629,63 @@ public void setSessionState(String sessionId, byte[] sessionState) {
     @Override
     public void close() {
         asyncClient.close();
+
+        if (messageProcessor.get() != null && !messageProcessor.get().isDisposed()) {
+            messageProcessor.get().dispose();
+        }
+
+        Disposable activeSubscription = messageProcessorSubscription.get();
+        if (activeSubscription != null && !activeSubscription.isDisposed()) {
+            activeSubscription.dispose();
+        }
     }
 
     /**
-     * Given an {@code emitter}, queues that work in {@link SynchronousMessageSubscriber}. If the synchronous job has
-     * not been created, will initialise it.
+     * Given an {@code emitter}, creates a {@link EmitterProcessor} to receive messages from Service Bus. If the
+     * message processor has not been created, will initialise it.
      */
     private void queueWork(int maximumMessageCount, Duration maxWaitTime,
         FluxSink<ServiceBusReceivedMessageContext> emitter) {
-        final long id = idGenerator.getAndIncrement();
-        final SynchronousReceiveWork work = new SynchronousReceiveWork(id, maximumMessageCount, maxWaitTime,
-            emitter);
-        final SynchronousMessageSubscriber syncSubscriber = new SynchronousMessageSubscriber(work);
 
-        logger.info("[{}]: Started synchronous message subscriber.", id);
-        asyncClient.receive(DEFAULT_RECEIVE_OPTIONS).subscribeWith(syncSubscriber);
+        if (messageProcessor.get() != null && messageProcessor.get().isDisposed()) {
+            logger.error("[{}]: Can not receive messaged because client is closed.", asyncClient.getEntityPath());
+            return;
+        }
+
+        if (messageProcessor.get() == null) {
+            logger.info("[{}]: Creating EmitterProcessor message processor for entity.", asyncClient.getEntityPath());
+
+            EmitterProcessor<ServiceBusReceivedMessageContext> newProcessor = asyncClient.receive(DEFAULT_RECEIVE_OPTIONS)
+                .subscribeWith(EmitterProcessor.create(false));
+            // if some other thread have come in between, we will not reset new processor
+            if (!messageProcessor.compareAndSet(null, newProcessor)) {
+                newProcessor.dispose();
+            }
+
+            logger.info("[{}]: Started EmitterProcessor message processor for entity.",
+                asyncClient.getEntityPath());
+        }
+
+            logger.info("[{}]: Subscribing EmitterProcessor message processor for entity.",
+                asyncClient.getEntityPath());
+
+        Disposable newSubscription = messageProcessor.get()
+            .take(maximumMessageCount)
+            .timeout(maxWaitTime)
```

Comment:
`timeout()` is not the right operator here. The `maxWaitTime` param specifies the maximum time to wait before receiving all messages upto `maximumMessageCount`. `timeout()` operator will reset the timer after each message is received. Also, if there is a timeout, it will throw an exception which is not what should happen. Instead, it should return all messages until the timeout occurred.